src/main/kotlin/org/elm/lang/core/types/TypeReplacement.kt

```diff
@@ -77,38 +71,63 @@ class TypeReplacement(
          * when calling functions.
          */
         fun flexify(ty: Ty): Ty {
-            if (ty.allVars(true).none()) return ty
-            return TypeReplacement(emptyMap(), freshen = false, varsToRemainRigid = emptyList(), freeze = false, keepRecordsMutable = false).replace(ty)
+            return TypeReplacement(emptyMap(), freshen = false, varsToRemainRigid = emptyList(), keepRecordsMutable = false).replace(ty)
         }
 
         /**
          * Freeze any fields references present in records in a [ty].
          *
-         * Use this to prevent cached values from being altered.
+         * Use this to prevent cached values from being altered. Note that this is an in-place
+         * change rather than a copy.
          */
-        fun freeze(ty: Ty): Ty {
-            if (ty.traverse(true).none { it.isUnfrozenRecord() }) return ty
-            return TypeReplacement(emptyMap(), freshen = false, varsToRemainRigid = emptyList(), freeze = true, keepRecordsMutable = false).replace(ty)
-        }
-
-        private fun tyWouldChange(ty: Ty, freeze: Boolean): Boolean {
-            return ty.traverse(true).any {
-                it is TyVar || freeze && it.isUnfrozenRecord()
+        fun freeze(ty: Ty) {
+            when (ty) {
+                is TyVar -> Unit
+                is TyTuple -> ty.types.forEach { freeze(it) }
+                is TyRecord -> {
+                    (ty.baseTy as? TyRecord)?.fieldReferences?.freeze()
+                    ty.fields.values.forEach { freeze(it) }
+                    ty.fieldReferences.freeze()
+                }
+                is MutableTyRecord -> {
+                    (ty.baseTy as? TyRecord)?.fieldReferences?.freeze()
+                    ty.fields.values.forEach { freeze(it) }
+                    ty.fieldReferences.freeze()
+                }
+                is TyUnion -> ty.parameters.forEach { freeze(it) }
+                is TyFunction -> {
+                    freeze(ty.ret)
+                    ty.parameters.forEach { freeze(it) }
+                }
+                is TyUnit -> Unit
+                is TyUnknown -> Unit
+                TyInProgressBinding -> Unit
             }
+            ty.alias?.parameters?.forEach { freeze(it) }
         }
+    }
 
-        private fun Ty.isUnfrozenRecord() = when (this) {
-            is TyRecord -> !fieldReferences.frozen
-            is MutableTyRecord -> !fieldReferences.frozen
-            else -> false
-        }
+    private fun wouldChange(ty: Ty): Boolean {
```

Comment:
Because this is now a member method, I experimented with caching the result in an `IdentityHashMap` to avoid repeated traversal when walking down the ty for replacement. This didn't have any performance impact. The cache was cheap to look up, but was almost always a miss, since tys are generally shallow.